### PR TITLE
[platform_view] iOSP platformView composition optimize.

### PR DESCRIPTION
**Pull request by cyanglaz**
Optimize the performance of the platform view composition.
When recomposition is required, we previously remove all the platform view related views (intercepting view, platform view) and re-add them for each frame. This is mentioned in https://github.com/flutter/flutter/issues/23793.
This PR enhanced the performance by re-arrange the existing UIViews and only add the newly created Views.

As a result, it would also fix https://github.com/flutter/flutter/issues/29427

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 1]
```diff
@@ -227,20 +227,34 @@
...
   UIView* flutter_view = flutter_view_.get();
 
-  // This can be more efficient, instead of removing all views and then re-attaching them,
-  // we should only remove the views that has been completly removed from the layer tree, and
-  // reorder the views using UIView's bringSubviewToFront.
-  // TODO(amirh): make this more efficient.
-  // https://github.com/flutter/flutter/issues/23793
-  for (UIView* sub_view in [flutter_view subviews]) {
-    [sub_view removeFromSuperview];
+  std::unordered_set<int64_t> composition_order_set;
```
Comment (amirh) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: Let's extract this to a `DetachUnusedLayers` method.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 2]
```diff
@@ -429,6 +443,7 @@ - (void)touchesBegan:(NSSet*)touches withEvent:(UIEvent*)event {
 
 - (void)touchesMoved:(NSSet*)touches withEvent:(UIEvent*)event {
   [_flutterView touchesMoved:touches withEvent:event];
+  self.state = UIGestureRecognizerStateChanged;
```
Comment (amirh) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: Why are we doing this?

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 3]
```diff
@@ -227,20 +227,34 @@
...
   }
 
+  composition_order_set.clear();
   active_composition_order_.clear();
+
   for (size_t i = 0; i < composition_order_.size(); i++) {
     int view_id = composition_order_[i];
-    [flutter_view addSubview:touch_interceptors_[view_id].get()];
-    [flutter_view addSubview:overlays_[view_id]->overlay_view.get()];
+    if (touch_interceptors_[view_id].get().superview == flutter_view) {
```
Comment (amirh) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: Let's use the same condition for both the overlay and the interceptor.
They should always be attached/detached together, we can even assert that the superview for both is the same just before the condition.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 2 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: oops. forgot to remove this.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 1 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: done

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 3 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: done

**Comment by cyanglaz** [comment]
@amirh Fixed review comments, please give it a second pass.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h [hunk 4]
```diff
@@ -65,7 +65,9 @@ class FlutterPlatformViewsController {
   // Discards all platform views instances and auxiliary resources.
   void Reset();
 
-  bool SubmitFrame(bool gl_rendering,
+    void extracted();
```
Comment (amirh) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h]: what's that?

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h [hunk 5]
```diff
@@ -97,7 +99,8 @@ class FlutterPlatformViewsController {
   void OnDispose(FlutterMethodCall* call, FlutterResult& result);
   void OnAcceptGesture(FlutterMethodCall* call, FlutterResult& result);
   void OnRejectGesture(FlutterMethodCall* call, FlutterResult& result);
-
+    
+    void DetachUnusedLayers();
```
Comment (amirh) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h]: indentation

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 6]
```diff
@@ -225,29 +225,44 @@
...
+  }
+
+  for (int64_t view_id : active_composition_order_) {
+    if (composition_order_set.find(view_id) == composition_order_set.end()) {
+      [touch_interceptors_[view_id].get() removeFromSuperview];
+      [overlays_[view_id]->overlay_view.get() removeFromSuperview];
+    }
+  }
+
+  composition_order_set.clear();
```
Comment (amirh) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: That's probably unnecessary, it's about to be removed from the stack.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 7]
```diff
@@ -225,29 +225,44 @@
...
-  for (UIView* sub_view in [flutter_view subviews]) {
-    [sub_view removeFromSuperview];
-  }
-
-  active_composition_order_.clear();
   for (size_t i = 0; i < composition_order_.size(); i++) {
     int view_id = composition_order_[i];
-    [flutter_view addSubview:touch_interceptors_[view_id].get()];
-    [flutter_view addSubview:overlays_[view_id]->overlay_view.get()];
+    if (touch_interceptors_[view_id].get().superview == flutter_view) {
```
Comment (amirh) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: You can add an assert that `touch_interceptors_[view_id].get().superview` is the same as `flutter_view bringSubviewToFront:overlays_[view_id]->overlay_view.get().superview` (also probably would be cleaner to have 2 local variables for the 2 views in this loop)

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h (hunk 4 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h]: sorry, this is prob caused by xcode auto extraction tool. I am surprised it didn't come up when compile.

Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h]: oh it is in header.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 6 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: done

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h (hunk 5 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h]: done

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 3 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: done

**Comment by cyanglaz** [comment]
@amirh Fixed all the review comments.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 8]
```diff
@@ -225,29 +226,46 @@
...
-  }
-
-  active_composition_order_.clear();
   for (size_t i = 0; i < composition_order_.size(); i++) {
     int view_id = composition_order_[i];
-    [flutter_view addSubview:touch_interceptors_[view_id].get()];
-    [flutter_view addSubview:overlays_[view_id]->overlay_view.get()];
+      UIView *intercepter = touch_interceptors_[view_id].get();
+      UIView *overlay = overlays_[view_id]->overlay_view;
+      assert(intercepter.superview  == overlay.superview);
```
Comment (amirh) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: I think we usually use `NSAssert` in ObjC code? (@cbracken is that correct?)

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 9]
```diff
@@ -225,29 +226,46 @@
...
-  for (UIView* sub_view in [flutter_view subviews]) {
-    [sub_view removeFromSuperview];
-  }
-
-  active_composition_order_.clear();
   for (size_t i = 0; i < composition_order_.size(); i++) {
     int view_id = composition_order_[i];
-    [flutter_view addSubview:touch_interceptors_[view_id].get()];
-    [flutter_view addSubview:overlays_[view_id]->overlay_view.get()];
+      UIView *intercepter = touch_interceptors_[view_id].get();
```
Comment (amirh) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: indentation

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h [hunk 10]
```diff
@@ -64,7 +64,7 @@ class FlutterPlatformViewsController {
 
   // Discards all platform views instances and auxiliary resources.
   void Reset();
-
+
```
Comment (amirh) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h]: revert

Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h]: done

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 9 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: done

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 8 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: done. Used FML_CHECK after checking with @cbracken offline.

**Pull request merged**